File: pepcoding/week5/coin.cpp
#include<iostream>
#include<string>
#include<vector>
using namespace std;


//creating permumation for coins using inifinte coins

int permuInfi(vector<int> &arr, int tar, string ans)
{
    if (tar == 0)
    {
        cout << ans << endl;
        return 1;
    }
    int count = 0;
    for (int i = 0; i < arr.size(); i++)
    {
        if (tar - arr[i] >= 0)
            count += permuInfi(arr, tar - arr[i], ans + to_string(arr[i]) + " ");
    }
    return count;
}




int combiInfi(vector<int> &arr, int idx, int tar, string ans)
{
    if (tar == 0)
    {
        cout << ans << endl;
        return 1;
    }
    int count = 0;
    for (int i = idx; i < arr.size(); i++)
    {
        if (tar - arr[i] >= 0)
            count += combiInfi(arr, i, tar - arr[i], ans + to_string(arr[i]) + " ");
    }
    return count;
}


int combiWithoutRepe(vector<int> &arr, int idx, int tar, string ans)
{
    if (tar == 0)
    {
        cout << ans << endl;
        return 1;
    }
    int count = 0;

    for (int i = idx; i < arr.size(); i++)
    {
        if (tar - arr[i] >= 0)
            count += combiWithoutRepe(arr, i + 1, tar - arr[i], ans + to_string(arr[i]) + " ");
    }
    return count;
}

int permuWithoutRepe(vector<int> &arr, vector<bool> &vis, int tar, string ans)
{
    if (tar == 0)
    {
        cout << ans << endl;
        return 1;
    }
    int count = 0;
    for (int i = 0; i < arr.size(); i++)
    {
        if (!vis[i] && tar - arr[i] >= 0)
        {
            vis[i] = true;
            count += permuWithoutRepe(arr, vis, tar - arr[i], ans + to_string(arr[i]) + " ");
            vis[i] = false;
        }
    }
    return count;
}



void basicQues()
{
    
     vector<int> arr{2, 3, 5, 7};
       cout << permuInfi(arr, 10, "") << endl;
   
     cout<<combiInfi(arr,0,10,"")<<endl;

     //cout<<combiusngSinglecoin(arr,0,10,"")<<endl;
    cout << combiWithoutRepe(arr, 0, 10, "") << endl;

    vector<bool>vis(arr.size(),false);

   cout << permuWithoutRepe(arr, vis, 10, "") << endl;



}
void solve()
{
    basicQues();
}
int main()
{
    solve();
}

File: pepcoding/week5/coin.java
import java.util.Scanner;
class coin
{
    public static int perumuationInfinteCoin(int arr[],int target,String ans)
    {
        if(target==0)
        {
            System.out.println(ans);
            return 1;
        }

        int count=0;

        for(int i=0;i<arr.length;i++)
        {
            if(target-arr[i]>0)
            {
                count+=perumuationInfinteCoin(arr,target-arr[i],ans+arr[i]);
            }
           
        }
         return count;
    }


    public static int combinationInfinteCoin(int arr[]int idx,int target,String ans)
    {
        if(target==0)
        {
            System.out.println(ans);
            return 1;
        }

        int count=0;

        for(int i=idx;i<arr.length;i++)
        {
            if(target-arr[i]>0)
            {
                count+=combinationInfinteCoin(arr,i,target-arr[i],ans+arr[i]);
            }
           
        }
         return count;
    }
    

    public static int combinationSingleCoin(int arr[]int idx,int target,String ans)
    {
        if(target==0)
        {
            System.out.println(ans);
            return 1;
        }

        int count=0;

        for(int i=idx;i<arr.length;i++)
        {
            if(target-arr[i]>0)
            {
                count+=combinationInfinteCoin(arr,i,target-arr[i],ans+arr[i]);
            }
           
        }
         return count;
    }
    public static  void  basicQues()
    {
        int arr[]=new int[]{2,3,5,7};
        System.out.println( perumuationInfinteCoin(arr,10,""));
         System.out.println( combinationInfinteCoin(arr,0,10,""));

    }
    public static void solve()
    {
        basicQues();
    }
    public static void main(String args[])
    {
        solve();

    }

}